File: js/cli/src/commands/e2e.ts
import { Command } from 'commander';
import {
  DaemonClient,
  ensurePersistedE2EConfig,
  getAlias,
  getIdentity,
  setE2EConfig,
} from '@quadra-a/runtime';
import type { LocalE2EConfig } from '@quadra-a/protocol';
import { error, info, printHeader, printKeyValue, success, warn } from '../ui.js';

function sessionPeerDid(sessionKey: string): string | undefined {
  const delimiter = sessionKey.lastIndexOf(':');
  if (delimiter <= 0) {
    return undefined;
  }

  return sessionKey.slice(0, delimiter) || undefined;
}

function collectResetPeers(sessionKeys: string[], onlyPeer?: string): string[] {
  const peers = new Set<string>();
  const exactPrefix = onlyPeer ? `${onlyPeer}:` : undefined;

  for (const sessionKey of sessionKeys) {
    if (exactPrefix && !sessionKey.startsWith(exactPrefix)) {
      continue;
    }

    const peerDid = sessionPeerDid(sessionKey);
    if (peerDid) {
      peers.add(peerDid);
    }
  }

  return [...peers].sort();
}

function isValidLocalE2EConfig(config: LocalE2EConfig | undefined): config is LocalE2EConfig {
  return Boolean(config?.currentDeviceId && config.devices?.[config.currentDeviceId]);
}

export function registerE2ECommand(program: Command): void {
  const e2e = program
    .command('e2e')
    .description('Manage local E2E session state');

  e2e
    .command('status')
    .description('Show E2E session status')
    .option('--json', 'Output as JSON')
    .action(async (options) => {
      try {
        const identity = getIdentity();
        if (!identity) {
          error('No identity found. Run "a4 listen" to initialize.');
          process.exit(1);
        }

        const { e2eConfig } = await ensurePersistedE2EConfig(identity);
        const currentDevice = e2eConfig.devices[e2eConfig.currentDeviceId];
        if (!currentDevice) {
          error('Current device not found in E2E config.');
          process.exit(1);
        }

        const sessions = Object.keys(currentDevice.sessions ?? {}).map((key) => ({
          key,
          peer: sessionPeerDid(key) ?? 'unknown',
        }));

        if (options.json) {
          console.log(JSON.stringify({
            deviceId: e2eConfig.currentDeviceId,
            sessionCount: sessions.length,
            sessions,
          }, null, 2));
          return;
        }

        printHeader('E2E Status');
        printKeyValue('Device ID', e2eConfig.currentDeviceId);
        printKeyValue('Sessions', String(sessions.length));

        if (sessions.length > 0) {
          console.log();
          for (const session of sessions) {
            printKeyValue(session.peer, session.key);
          }
        }
      } catch (err) {
        error(`Failed to show E2E status: ${(err as Error).message}`);
        process.exit(1);
      }
    });

  e2e
    .command('reset [peer]')
    .description('Clear E2E sessions for one peer or all peers')
    .action(async (peer?: string) => {
      try {
        const identity = getIdentity();
        if (!identity) {
          error('No identity found. Run "a4 listen" to initialize.');
          process.exit(1);
        }

        const resolvedPeer = peer ? (getAlias(peer) ?? peer) : undefined;
        const { e2eConfig } = await ensurePersistedE2EConfig(identity);

        if (!isValidLocalE2EConfig(e2eConfig)) {
          error('Invalid E2E configuration. Run "a4 listen" to reinitialize.');
          process.exit(1);
        }

        const currentDeviceId = e2eConfig.currentDeviceId;
        const currentDevice = e2eConfig.devices[currentDeviceId];
        const currentSessions = currentDevice.sessions ?? {};
        const sessionKeys = Object.keys(currentSessions);
        const peersToNotify = collectResetPeers(sessionKeys, resolvedPeer);

        const nextSessions = resolvedPeer
          ? Object.fromEntries(
            Object.entries(currentSessions).filter(([sessionKey]) => !sessionKey.startsWith(`${resolvedPeer}:`)),
          )
          : {};
        const removedCount = sessionKeys.length - Object.keys(nextSessions).length;

        setE2EConfig({
          ...e2eConfig,
          devices: {
            ...e2eConfig.devices,
            [currentDeviceId]: {
              ...currentDevice,
              sessions: nextSessions,
            },
          },
        });

        if (resolvedPeer) {
          success(`Cleared ${removedCount} session(s) for peer ${resolvedPeer}`);
        } else {
          success(`Cleared all ${removedCount} session(s)`);
        }

        const client = new DaemonClient();
        if (await client.isDaemonRunning()) {
          try {
            await client.send('reload-e2e', {});
            info('Daemon E2E config reloaded.');
          } catch {
            warn('Daemon not responding. Changes will apply on next daemon start.');
          }

          if (peersToNotify.length > 0) {
            try {
              const result = await client.send<{
                notified?: string[];
                failed?: Array<{ peer: string; error: string }>;
              }>('e2e-reset-notify', { peers: peersToNotify });
              const notified = result.notified?.length ?? 0;
              const failed = result.failed?.length ?? 0;
              info(
                `Sent session reset notification to ${notified} peer(s)${
                  failed > 0 ? ` (${failed} failed)` : ''
                }.`,
              );
            } catch {
              warn('Could not notify peer(s) about the reset. Remote sessions will recover on next decrypt failure.');
            }
          }
        } else if (peersToNotify.length > 0) {
          warn('Daemon is not running, so peers were not notified about the reset. Remote sessions will recover on next decrypt failure.');
        }
      } catch (err) {
        error(`Failed to reset E2E sessions: ${(err as Error).message}`);
        process.exit(1);
      }
    });
}

File: rust/cli-rs/src/commands/e2e.rs
use anyhow::{Context, Result};
use serde_json::json;
use std::collections::BTreeSet;

use crate::config::load_config;
use crate::daemon::{daemon_socket_path, DaemonClient};
use crate::e2e_state::with_locked_config_transaction;

pub struct E2eStatusOptions {
    pub json: bool,
}

pub struct E2eResetOptions {
    pub peer_did: Option<String>,
}

fn session_peer_did(session_key: &str) -> Option<&str> {
    session_key
        .rsplit_once(':')
        .map(|(peer_did, _)| peer_did)
        .filter(|peer_did| !peer_did.is_empty())
}

fn collect_reset_peers<'a, I>(session_keys: I, only_peer: Option<&str>) -> Vec<String>
where
    I: IntoIterator<Item = &'a str>,
{
    let mut peers = BTreeSet::new();
    let exact_prefix = only_peer.map(|peer| format!("{}:", peer));

    for session_key in session_keys {
        let Some(peer_did) = session_peer_did(session_key) else {
            continue;
        };
        if let Some(prefix) = &exact_prefix {
            if !session_key.starts_with(prefix) {
                continue;
            }
        }
        peers.insert(peer_did.to_string());
    }

    peers.into_iter().collect()
}

pub async fn e2e_status(opts: E2eStatusOptions) -> Result<()> {
    let config = load_config()?;

    let e2e = config
        .e2e
        .as_ref()
        .context("No E2E configuration found. Run 'a4 listen' to initialize.")?;
    if !e2e.is_valid() {
        anyhow::bail!("Invalid E2E configuration. Run 'a4 listen' to reinitialize.");
    }

    let device_id = &e2e.current_device_id;
    let device = e2e
        .devices
        .get(device_id)
        .context("Current device not found in E2E config")?;

    let session_count = device.sessions.len();
    let sessions: Vec<_> = device
        .sessions
        .keys()
        .map(|k| {
            let parts: Vec<&str> = k.split(':').collect();
            json!({
                "key": k,
                "peer": parts.get(0).unwrap_or(&"unknown"),
            })
        })
        .collect();

    if opts.json {
        let output = json!({
            "deviceId": device_id,
            "sessionCount": session_count,
            "sessions": sessions,
        });
        println!("{}", serde_json::to_string_pretty(&output)?);
    } else {
        println!("E2E Status:");
        println!("  Device ID: {}", device_id);
        println!("  Sessions: {}", session_count);
        if !sessions.is_empty() {
            println!("\nActive sessions:");
            for session in sessions {
                println!(
                    "  - {} (key: {})",
                    session["peer"].as_str().unwrap_or("unknown"),
                    session["key"].as_str().unwrap_or("unknown")
                );
            }
        }
    }

    Ok(())
}

pub async fn e2e_reset(opts: E2eResetOptions) -> Result<()> {
    let config = load_config()?;

    // Resolve alias to DID if needed
    let peer_did = opts.peer_did.map(|input| {
        crate::commands::alias::resolve_did(&input, &config).unwrap_or(input)
    });
    let peer_did_for_reset = peer_did.clone();
    let (reset_result, _) = with_locked_config_transaction(|mut config| async move {
        let e2e = config
            .e2e
            .as_mut()
            .context("No E2E configuration found. Run 'a4 listen' to initialize.")?;
        if !e2e.is_valid() {
            anyhow::bail!("Invalid E2E configuration. Run 'a4 listen' to reinitialize.");
        }

        let device_id = &e2e.current_device_id.clone();
        let device = e2e
            .devices
            .get_mut(device_id)
            .context("Current device not found in E2E config")?;

        let before = device.sessions.len();
        let peers_to_notify = collect_reset_peers(
            device.sessions.keys().map(String::as_str),
            peer_did_for_reset.as_deref(),
        );

        let removed = if let Some(peer_did) = &peer_did_for_reset {
            let session_prefix = format!("{}:", peer_did);
            let before_peer = device.sessions.len();
            device.sessions.retain(|key, _| !key.starts_with(&session_prefix));
            before_peer - device.sessions.len()
        } else {
            let removed = device.sessions.len();
            device.sessions.clear();
            removed
        };

        Ok(((before, removed, peers_to_notify), config))
    }).await?;
    let (before, removed, peers_to_notify) = reset_result;

    if let Some(peer_did) = &peer_did {
        println!("Cleared {} session(s) for peer {}", removed, peer_did);
    } else {
        println!("Cleared all {} session(s)", before);
    }

    // Notify daemon to reload E2E config if running
    let socket_path = daemon_socket_path();
    let client = DaemonClient::new(&socket_path);
    if client.is_running().await {
        match client.send_command("reload-e2e", serde_json::json!({})).await {
            Ok(_) => {
                println!("\nDaemon E2E config reloaded.");
            }
            Err(_) => {
                println!("\nNote: Daemon not responding. Changes will apply on next daemon start.");
            }
        }

        if !peers_to_notify.is_empty() {
            match client
                .send_command("e2e-reset-notify", json!({ "peers": peers_to_notify }))
                .await
            {
                Ok(result) => {
                    let notified = result
                        .get("notified")
                        .and_then(|value| value.as_array())
                        .map(|value| value.len())
                        .unwrap_or(0);
                    let failed = result
                        .get("failed")
                        .and_then(|value| value.as_array())
                        .map(|value| value.len())
                        .unwrap_or(0);
                    println!(
                        "Sent session reset notification to {} peer(s){}.",
                        notified,
                        if failed > 0 {
                            format!(" ({} failed)", failed)
                        } else {
                            String::new()
                        }
                    );
                }
                Err(_) => {
                    println!(
                        "Note: Could not notify peer(s) about the reset. Remote sessions will recover on next decrypt failure."
                    );
                }
            }
        }
    } else if !peers_to_notify.is_empty() {
        println!(
            "\nNote: Daemon is not running, so peers were not notified about the reset. Remote sessions will recover on next decrypt failure."
        );
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::collect_reset_peers;

    #[test]
    fn collect_reset_peers_uses_exact_peer_prefixes() {
        let session_keys = vec![
            "did:agent:zAlice:device-1",
            "did:agent:zAliceExtra:device-2",
            "did:agent:zBob:device-3",
        ];

        assert_eq!(
            collect_reset_peers(session_keys.iter().copied(), Some("did:agent:zAlice")),
            vec!["did:agent:zAlice".to_string()]
        );
    }

    #[test]
    fn collect_reset_peers_dedupes_all_matching_peers() {
        let session_keys = vec![
            "did:agent:zAlice:device-1",
            "did:agent:zAlice:device-2",
            "did:agent:zBob:device-3",
        ];

        assert_eq!(
            collect_reset_peers(session_keys.iter().copied(), None),
            vec![
                "did:agent:zAlice".to_string(),
                "did:agent:zBob".to_string(),
            ]
        );
    }
}
